### PR TITLE
xmonad: document recompilation issue

### LINKED ISSUE #1895: xmonad recompilation makes `home-manager switch` fail
I was going to post this, but I found the following comment on a different issue that explains the problem.

> Somewhat related, more of a warning. If you have a xmonad configuration that includes a binding for both recompiling and restarting xmonad, it will override the nix managed xmonad symlink with the new binary. This will cause later nix rebuilds to fail because the file already exists.

> _Originally posted by @MasseR in https://github.com/nix-community/home-manager/issues/1891#issuecomment-810014778_

### PR DESCRIPTION
**Pull request by berbiche**
### Description

- xmonad: document breakage of recompilation
- news: add warning about xmonad recompilation

Whether this fixes #1895 is up to the author.

cc: @dnaq @markusscherer @woffs

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all`.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [ ] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/misc/news.nix [hunk 1]
```diff
@@ -2029,6 +2029,25 @@ in
           A new module is available: 'programs.foot'.
         '';
       }
+
+      {
+        time = "2021-05-19T01:26:16+00:00";
+        message = ''
+          The xmonad module now compiles the configuration beforing
```
Comment (sumnerevans) [inline modules/misc/news.nix:2036]:
```suggestion
          The xmonad module now compiles the configuration before
```

File: modules/misc/news.nix [hunk 2]
```diff
@@ -2029,6 +2029,25 @@ in
...
+          linking the binary to the place xmonad expects to find
+          the compiled configuration (the binary).
```
Comment (sumnerevans) [inline modules/misc/news.nix:2038]: Maybe I don't understand xmoad, but I find this description to be quite confusing.

I think my biggest confusion is what is the binary? Is it the config? Is it the xmonad executable? Maybe splitting this into two sentences would help.

File: modules/misc/news.nix [hunk 3]
```diff
@@ -2029,6 +2029,25 @@ in
...
+
+          This breaks recompilation of xmonad (i.e. the 'q' binding or
+          'xmonad --recompile').
+
+          If this behavior is undesirable, instead of using the option
+          'xsession.windowManager.xmonad.config', you may set the path
+          of the configuration file directory with
+          'home.file.".xmonad/config.hs".text = "content of the file"'
+          or 'home.file.".xmonad/config.hs".source = ./path-to-config'.
```
Comment (sumnerevans) [inline modules/misc/news.nix:2047]: Maybe it's just late, but I think that this sentence is a bit hard to parse as-is. Maybe something along these lines would be more clear?

```suggestion
          If this behavior is undesirable, do not use the
          'xsession.windowManager.xmonad.config' option. Instead, set the
          contents of the configuration file with
          'home.file.".xmonad/config.hs".text = "content of the file"'
          or 'home.file.".xmonad/config.hs".source = ./path-to-config'.
```

**Review by sumnerevans — CHANGES_REQUESTED** [review]
I don't know much about xmonad, so some of my comments may be totally wrong/make no sense.

File: modules/misc/news.nix (hunk 2 above)
Comment (fricklerhandwerk) [reply modules/misc/news.nix:2038]: `xmonad` inherits from the [suckless](https://suckless.org/) philosophy the trait that its configuration is part of the program, therefore written in the program's implementation language (here: Haskell), and *compiled into the binary*.  In that sense, the sentence is semantically correct, but a bit hard to untangle.

File: modules/misc/news.nix [hunk 4]
```diff
@@ -2030,6 +2030,7 @@ in
         '';
       }
 
+
```
Comment (sumnerevans) [inline modules/misc/news.nix:2033]: nit: please remove this. Something happened with the merge I think...

**Review by sumnerevans — CHANGES_REQUESTED** [review]
Also, please fix up the commit history to match the contributing guidelines.
